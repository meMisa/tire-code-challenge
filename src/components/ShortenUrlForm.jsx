/* eslint no-unused-vars: 1 */

import React, { useCallback, useRef } from 'react';
import shortenUrlApis from 'apis/shortenUrlApis';
// actions
import { URL_TO_SHOURTEN, URL, SHORTEN_COPY_URL } from 'constants/texts';
import { changeUrlValue,
    fetchShortenUrlRequest,
    fetchShortenUrlSuccess,
    fetchShortenUrlError } from './actions/shortenUrlActions';
// reducer
import { initialState, reducer } from './reducers/shortenReducer';
import 'styles/ShourtenUrlForm.css';

const ShortenUrlForm = () => {
    /** **************************************************** STATE  ****************************************** */
    const [state, dispatch] = React.useReducer(reducer, initialState);
    const shortUrlRef = useRef(null);
    /** **************************************************** FUNCTIONS  ****************************************** */
    const onChange = useCallback((e) => {
        dispatch(changeUrlValue({ value: e.target.value }));
    }, []);

    const copyToClipboard = useCallback((link) => {
        navigator.clipboard.writeText(link);
    }, []);
    const onSubmit = useCallback((e) => {
        e.preventDefault();
        if (!state.url || state.loading) return;

        dispatch(fetchShortenUrlRequest({ loading: true }));

        shortenUrlApis.postShortenUrl({ payload: { long_url: state.url } })
            .then((data) => {
                dispatch(fetchShortenUrlSuccess({ data }));
                copyToClipboard(data.link);
            })
            .catch((error) => {
                dispatch(fetchShortenUrlError(error));
            });

    }, [state.url, state.loading]);

    return (
        <div className="shourten-form-container">
            <form onSubmit={onSubmit}>
                <label htmlFor="shorten" data-testid="url-label">
                    {URL}
                    <input
                        className="url shorten-url-form-elements-common"
                        placeholder={URL_TO_SHOURTEN}
                        id="shorten"
                        type="url"
                        value={state.url}
                        onChange={onChange}
                    />
                </label>
                <div className="submit-wrapper">
                    <button
                        className={`shorten-url-form-elements-common ${state.loading ? 'disable-button' : ''}`}
                        type="submit"
                        data-testid="submit-button"
                    >
                        {SHORTEN_COPY_URL}
                    </button>
                </div>
                <div className="short-url" ref={shortUrlRef}>
                    {state.shortUrl}
                </div>
            </form>
        </div>
    );
};

export default ShortenUrlForm;
