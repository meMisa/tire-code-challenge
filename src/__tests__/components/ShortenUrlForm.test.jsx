import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

import ShortenUrlForm from 'components/ShortenUrlForm';

afterEach(() => {
    cleanup();
});

describe('Test', () => {
    test('should render ShortenUrlForm component correctly', () => {
        render(<ShortenUrlForm />);
        const ShortenUrlFormComponent = screen.getByTestId('url-label');
        expect(ShortenUrlFormComponent).toBeInTheDocument();
        expect(ShortenUrlFormComponent).toHaveTextContent('Url:');
    });
    test('should have got button submit', () => {
        render(<ShortenUrlForm />);
        const ShortenUrlFormComponent = screen.getByTestId('submit-button');
        expect(ShortenUrlFormComponent).toBeInTheDocument();
        expect(ShortenUrlFormComponent).toHaveTextContent('Shorten and copy URL');
    });
    test('match snapshot', () => {
        const tree = renderer.create(<ShortenUrlForm />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
