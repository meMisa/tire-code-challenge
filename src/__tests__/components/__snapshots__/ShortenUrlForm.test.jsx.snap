// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test match snapshot 1`] = `
<div
  className="shourten-form-container"
>
  <form
    onSubmit={[Function]}
  >
    <label
      data-testid="url-label"
      htmlFor="shorten"
    >
      Url:
      <input
        className="url shorten-url-form-elements-common"
        id="shorten"
        onChange={[Function]}
        placeholder="Url to shorten"
        type="url"
        value=""
      />
    </label>
    <div
      className="submit-wrapper"
    >
      <button
        className="shorten-url-form-elements-common "
        data-testid="submit-button"
        type="submit"
      >
        Shorten and copy URL
      </button>
    </div>
    <div
      className="short-url"
    >
      
    </div>
  </form>
</div>
`;
